oneview/resource_logical_enclosure.go

```diff
@@ -0,0 +1,377 @@
+		"logical_interconnect_update_mode":                 logicalEnclosure.Firmware.LogicalInterconnectUpdateMode,
+		"update_firmware_on_unmanaged_interconnect":        logicalEnclosure.Firmware.UpdateFirmwareOnUnmanagedInterconnect,
+		"validate_if_li_firmware_update_is_non_disruptive": logicalEnclosure.Firmware.ValidateIfLIFirmwareUpdateIsNonDisruptive,
+	})
+	d.Set("firmware", logicalEnclosureFirmware)
+	d.Set("ip_addressing_mode", logicalEnclosure.IpAddressingMode)
+	logicalEnclosureIpv4Ranges := make([]map[string]interface{}, 0, len(logicalEnclosure.Ipv4Ranges))
+	for _, logicalEnclosureIpv4Range := range logicalEnclosure.Ipv4Ranges {
+		dnsServerMap := make([]interface{}, len(logicalEnclosureIpv4Range.DnsServers))
+		for i, dnsServer := range logicalEnclosureIpv4Range.DnsServers {
+			dnsServerMap[i] = dnsServer
+		}
+		logicalEnclosureIpv4Ranges = append(logicalEnclosureIpv4Ranges, map[string]interface{}{
+			"dns_servers":  schema.NewSet(schema.HashString, dnsServerMap),
+			"domain":       logicalEnclosureIpv4Range.Domain,
+			"gateway":      logicalEnclosureIpv4Range.Gateway,
+			"ip_range_uri": logicalEnclosureIpv4Range.IpRangeUri,
+			"name":         logicalEnclosureIpv4Range.Name,
+			"subnet_mask":  logicalEnclosureIpv4Range.SubnetMask,
+		})
+	}
+	ipv4RangeCount := d.Get("ipv4_range.#").(int)
+	oneviewIpv4RangeCount := len(logicalEnclosureIpv4Ranges)
+	for i := 0; i < ipv4RangeCount; i++ {
+		currIpv4RangeName := d.Get("ipv4_range." + strconv.Itoa(i) + ".name").(string)
+		for j := 0; j < oneviewIpv4RangeCount; j++ {
+			if currIpv4RangeName == logicalEnclosureIpv4Ranges[j]["name"] && i <= j {
+				logicalEnclosureIpv4Ranges[i], logicalEnclosureIpv4Ranges[j] = logicalEnclosureIpv4Ranges[j], logicalEnclosureIpv4Ranges[i]
+			}
+		}
+	}
+	d.Set("ipv4_range", logicalEnclosureIpv4Ranges)
+	d.Set("logical_interconnect_uris", logicalEnclosure.LogicalInterconnectUris)
+	d.Set("modified", logicalEnclosure.Modified)
+	d.Set("name", logicalEnclosure.Name)
+	d.Set("power_mode", logicalEnclosure.PowerMode)
+	d.Set("scaling_state", logicalEnclosure.ScalingState)
+	d.Set("scopes_uri", logicalEnclosure.ScopesUri)
+	d.Set("status", logicalEnclosure.Status)
+	d.Set("type", logicalEnclosure.Type)
+	d.Set("uri", logicalEnclosure.URI.String())
+	return nil
+}
+
+func resourceLogicalEnclosureUpdate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	logicalEnclosure := ov.LogicalEnclosure{
+		Name: d.Get("name").(string),
+		AmbientTemperatureMode: d.Get("ambient_temperature_mode").(string),
+		DeleteFailed:           d.Get("delete_failed").(bool),
+		URI:                    utils.NewNstring(d.Get("uri").(string)),
+		PowerMode:              d.Get("power_mode").(string),
+		ScalingState:           d.Get("scaling_state").(string),
+		ScopesUri:              utils.NewNstring(d.Get("scopes_uri").(string)),
+		Type:                   d.Get("type").(string),
```

Comment:
Few fields in this PUT are optional, so using GetOk is preferred. Check the documentation once.

Reply:
Corrected that, optional fields are reading using the preferred way 'GetOk'